homeassistant/components/zwave_js/triggers/event.py

```diff
@@ -0,0 +1,193 @@
+"""Offer Z-Wave JS event listening automation trigger."""
+from __future__ import annotations
+
+import functools
+
+import voluptuous as vol
+from zwave_js_server.client import Client
+from zwave_js_server.model.node import Node
+
+from homeassistant.components.automation import (
+    AutomationActionType,
+    AutomationTriggerInfo,
+)
+from homeassistant.components.zwave_js.const import (
+    ATTR_CONFIG_ENTRY_ID,
+    ATTR_EVENT,
+    ATTR_EVENT_DATA,
+    ATTR_EVENT_SOURCE,
+    ATTR_NODE_ID,
+    ATTR_PARTIAL_DICT_MATCH,
+    DATA_CLIENT,
+    DOMAIN,
+)
+from homeassistant.components.zwave_js.helpers import (
+    async_get_node_from_device_id,
+    async_get_node_from_entity_id,
+    get_device_id,
+    get_home_and_node_id_from_device_id,
+)
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.const import ATTR_DEVICE_ID, ATTR_ENTITY_ID, CONF_PLATFORM
+from homeassistant.core import CALLBACK_TYPE, HassJob, HomeAssistant, callback
+from homeassistant.helpers import config_validation as cv, device_registry as dr
+from homeassistant.helpers.typing import ConfigType
+
+# Platform type should be <DOMAIN>.<SUBMODULE_NAME>
+PLATFORM_TYPE = f"{DOMAIN}.{__name__.rsplit('.', maxsplit=1)[-1]}"
+
+
+def validate_non_node_event_source(obj: dict) -> dict:
+    """Validate that a trigger for a non node event source has a config entry."""
+    if obj[ATTR_EVENT_SOURCE] != "node" and ATTR_CONFIG_ENTRY_ID in obj:
+        return obj
+    raise vol.Invalid(f"Non node event triggers must contain {ATTR_CONFIG_ENTRY_ID}.")
+
+
+TRIGGER_SCHEMA = vol.All(
+    cv.TRIGGER_BASE_SCHEMA.extend(
+        {
+            vol.Required(CONF_PLATFORM): PLATFORM_TYPE,
+            vol.Optional(ATTR_CONFIG_ENTRY_ID): str,
+            vol.Optional(ATTR_DEVICE_ID): vol.All(cv.ensure_list, [cv.string]),
+            vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
+            vol.Required(ATTR_EVENT_SOURCE): vol.In(["node", "controller", "driver"]),
+            vol.Required(ATTR_EVENT): cv.string,
+            vol.Optional(ATTR_EVENT_DATA): dict,
```

Comment:
Would be good to be able to validate the event data filter to match the event we're targeting.

Reply:
oof that adds a lot of extra typing requirements to the library (I'm assuming we would want to maintain the event to event data mapping in the lib). Probably a good idea but a decently large addition.

Reply:
see my other comment